### PR TITLE
Diminished Tuple Confusion

### PR DESCRIPTION
**Pull request by retronym**
- Better error message for ((a, b) => c): (Tuple2[A, B] => C)
- Better error message for (((a, b)) => c)

Review by @densh @soc

### CODE REVIEW

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala [hunk 1]
```diff
@@ -739,9 +740,17 @@ self =>
         case Typed(Ident(name), tpe) if tpe.isType => // get the ident!
           removeAsPlaceholder(name)
           makeParam(name.toTermName, tpe)
+        case build.SyntacticTuple(as) =>
+          val arity = as.length
+          syntaxError(tree.pos,
+            sm"""|not a legal formal parameter.
+                 |Note: Tuples cannot be directly destructured in method or function parameters.
+                 |      Either create a single parameter accepting the Tuple${as.length},
+                 |      or consider a pattern matching anonymous function: `{ case (param1, ..., param${arity}}) => ... }""", skipIt = false)
```
Comment (kiritsuku) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala]: `s/})/)/`. Furthermore, having the parameter names directly printed into the error message looks nicer.

**Comment by retronym** [comment]
@sschaef : I've been more specific with the parameter names (except if they were upper cases and unsuitable as variable patterns) and fixed the typo.

**Comment by densh** [comment]
LGTM. Nice to see Syntactics being used outside of quasiquotes.

File: src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala [hunk 2]
```diff
@@ -408,11 +408,28 @@ trait ContextErrors {
...
+        def issue(what: String) = {
+          val addendum: String = fun match {
+            case Function(params, _) if withTupleAddendum =>
+              val funArity = params.length
+              val example = analyzer.exampleTuplePattern(params map (_.name))
+              (pt baseType FunctionClass(1)) match {
+                case TypeRef(_, _, arg :: _) if arg.typeSymbol == TupleClass(funArity) && funArity > 1 =>
+                  sm"""|
+                       |Note: The expected type requires a one-argument function accepting a $funArity-Tuple.
+                       |      Consider a pattern matching anoynmous function, `{ case $example =>  ... }`"""
```
Comment (pavelpavlov) [inline src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala:421]: typo: anoynmous => anonymous

**Pull request merged**